R/germline.R

```diff
@@ -93,19 +76,58 @@ take_first_allele <- function(string) {
   unlist(strsplit(string, ","))[1]
 }
 
-generate_germline_sequence <- function(v_seq, j_seq) {
-  if (is.na(v_seq) || is.na(j_seq)) {
+generate_germline_sequence <- function(seq, v_ref, v_end, cdr3_start, cdr3_end, j_start) {
+  if (is.na(seq) || is.na(v_ref) ||
+    is.na(v_end) || is.na(cdr3_start) || is.na(cdr3_end) || is.na(j_start)) {
     return(NA)
   } else {
-    return(paste(v_seq, j_seq, sep = "..."))
+    cdr3_start %<>% as.numeric()
+    cdr3_end %<>% as.numeric()
+
+    if (v_end <= cdr3_start) {
+      v_part <- v_ref
+    } else {
+      # trim intersection of V and CDR3 from reference V gene
+      v_part <- stringr::str_sub(
+        v_ref, 1,
+        max(0, stringr::str_length(v_ref) - (v_end - cdr3_start))
+      )
+    }
+
+    cdr3_part <- paste(rep("n", cdr3_end - cdr3_start), collapse = "")
+
+    j_part <- stringr::str_sub(seq, max(cdr3_end, j_start))
+
+    paste0(v_part, cdr3_part, j_part) %>%
+      toupper() %>%
+      return()
   }
 }
 
-load_reference_sequences <- function(chain, species) {
+merge_reference_sequences <- function(data, chain, species, sample_name) {
   data(genesegments)
-  sequences_df <- GENE_SEGMENTS %>% filter(species == species)
-  sequences_df <- sequences_df[c("sequence", "allele_id")]
+  reference_df <- GENE_SEGMENTS %>% filter(species == species)
+  reference_df <- reference_df[c("sequence", "allele_id")]
   chain_letter <- stringr::str_sub(chain, -1)
-  colnames(sequences_df) <- c(paste0(chain_letter, ".sequence"), paste0(chain_letter, ".first.allele"))
-  return(sequences_df)
+  chain_seq_colname <- paste0(chain_letter, ".sequence")
+  chain_allele_colname <- paste0(chain_letter, ".first.allele")
+  colnames(reference_df) <- c(chain_seq_colname, chain_allele_colname)
+
+  # check for alleles in data that don't exist in the reference
+  alleles_in_data <- unique(data[[chain_allele_colname]])
+  alleles_in_ref <- unique(reference_df[[chain_allele_colname]])
+  missing_alleles <- alleles_in_data[!(alleles_in_data %in% alleles_in_ref)]
+  if (length(missing_alleles) > 0) {
+    warning(
```

Comment:
Please see the code style for long lines in warnings
https://style.tidyverse.org/syntax.html?q=string#long-lines